addons/mail/static/src/discuss/gif_picker/gif_picker.js

```diff
@@ -0,0 +1,219 @@
+/* @odoo-module */
+
+import { useOnBottomScrolled } from "@mail/utils/hooks";
+import { Component, onWillStart, useState } from "@odoo/owl";
+import { useService, useAutofocus } from "@web/core/utils/hooks";
+import { removeFromArrayWithPredicate } from "@mail/utils/arrays";
+import { useDebounced } from "@web/core/utils/timing";
+import { useStore } from "@mail/core/messaging_hook";
+
+/**
+ * @typedef {Object} TenorCategory
+ * @property {string} searchterm
+ * @property {string} path
+ * @property {string} image
+ * @property {string} name
+ */
+
+/**
+ * @typedef {Object} TenorMediaFormat
+ * @property {string} url
+ * @property {number} duration
+ * @property {string} preview
+ * @property {number[]} dims
+ * @property {number} size
+ */
+
+/**
+ * @typedef {Object} TenorGif
+ * @property {string} id
+ * @property {string} title
+ * @property {number} created
+ * @property {string} content_description
+ * @property {string} itemurl
+ * @property {string} url
+ * @property {string[]} tags
+ * @property {string[]} flags
+ * @property {boolean} hasaudio
+ * @property {{ tinygif: TenorMediaFormat }} media_formats
+ */
+
+/**
+ * @typedef {Object} Props
+ * @property {fucntion} onSelected Callback to use when the gif is selected
+ * @property {string} [className]
+ * @property {function} [close]
+ * @extends {Component<Props, Env>}
+ */
+export class GifPicker extends Component {
+    static template = "discuss.GifPicker";
+    static props = ["onSelected", "className?", "close?"];
+
+    async setup() {
+        this.rpc = useService("rpc");
+        this.store = useStore();
+        this.userService = useService("user");
+        useAutofocus();
+        useOnBottomScrolled(
+            "scroller",
+            () => {
+                if (!this.state.showCategories) {
+                    this.searchDebounced();
+                }
+            },
+            300
+        );
+        this.state = useState({
+            favorites: [],
+            searchTerm: "",
+            gifs: [],
+            next: "",
+            showCategories: true,
+            /** @type {TenorCategory[]} */
+            categories: [],
+            loadingGif: false,
+            loadingError: false,
+            evenGif: {
+                /** @type {TenorGif[]} */
+                gifs: [],
+                size: 0,
+            },
+            oddGif: {
+                /** @type {TenorGif[]} */
+                gifs: [],
+                size: 0,
+            },
+        });
+        this.searchDebounced = useDebounced(this.search, 200);
+        onWillStart(() => {
+            this.loadCategories();
+        });
+        if (!this.store.guest) {
+            onWillStart(() => {
+                this.loadFavorites();
+            });
```

Comment:
I made `loadCategories` and `loadFavorites` non-blocking, so that opening of gif picker looks faster

Reply:
+ `{ shadow: true }`